app/models/mhv_account.rb

```diff
@@ -29,60 +15,88 @@ class MhvAccount < ActiveRecord::Base
+  # "existing" even though they might have actually been created by us and a single uuid can track multiple
+  # different mhv_correlation_ids even though only 1 should ever be the active one.
+  scope :historic, -> { where(mhv_correlation_id: nil) }
+  scope :active, -> { where.not(mhv_correlation_id: nil) }
+
+  TERMS_AND_CONDITIONS_NAME = 'mhvac'
+  UPGRADABLE_ACCOUNT_LEVELS = [nil, 'Basic', 'Advanced'].freeze
+  INELIGIBLE_STATES = %i[
+    needs_identity_verification needs_ssn_resolution needs_va_patient
+    has_deactivated_mhv_ids has_multiple_active_mhv_ids
+    state_ineligible country_ineligible needs_terms_acceptance
+  ].freeze
+  PERSISTED_STATES = %i[registered upgraded register_failed upgrade_failed].freeze
+  ELIGIBLE_STATES = %i[existing eligible no_account].freeze
+  ALL_STATES = (%i[unknown] + INELIGIBLE_STATES + ELIGIBLE_STATES + PERSISTED_STATES).freeze
 
   after_initialize :setup
 
+  # rubocop:disable Metrics/BlockLength
   aasm(:account_state) do
     state :unknown, initial: true
-    state :needs_terms_acceptance, :existing, :ineligible, :registered, :upgraded, :register_failed, :upgrade_failed
+    state(*(INELIGIBLE_STATES + ELIGIBLE_STATES + PERSISTED_STATES))
 
+    # NOTE: This is eligibility for account creation or upgrade, not for access to services.
     event :check_eligibility do
-      transitions from: ALL_STATES, to: :existing, if: :preexisting_account?
-      transitions from: ALL_STATES, to: :ineligible, unless: :eligible?
-      transitions from: ALL_STATES, to: :upgraded, if: :previously_upgraded?
-      transitions from: ALL_STATES, to: :registered, if: :previously_registered?
-      transitions from: ALL_STATES, to: :unknown
+      transitions from: ALL_STATES, to: :needs_identity_verification, unless: :identity_proofed?
+      transitions from: ALL_STATES, to: :needs_ssn_resolution, if: :ssn_mismatch?
+      transitions from: ALL_STATES, to: :needs_va_patient, unless: :va_patient?
+      transitions from: ALL_STATES, to: :has_deactivated_mhv_ids, if: :deactivated_mhv_ids?
+      transitions from: ALL_STATES, to: :has_multiple_active_mhv_ids, if: :multiple_active_mhv_ids?
+      transitions from: ALL_STATES, to: :needs_terms_acceptance, if: :requires_terms_acceptance?
+      transitions from: ALL_STATES, to: :eligible
     end
 
-    event :check_terms_acceptance do
-      transitions from: ALL_STATES - %i[existing ineligible],
-                  to: :needs_terms_acceptance, unless: :terms_and_conditions_accepted?
+    # FIXME: revisit these in the future and see if they can be cleaned up
+    # in the future might need to consider downgrades from upgrade, if account level can be changed.
+    event :check_account_state do
+      transitions from: %i[eligible], to: :no_account, unless: :exists?
+      transitions from: %i[eligible], to: :upgraded, if: :previously_registered_somehow_upgraded?
+      transitions from: %i[eligible], to: :registered, if: :previously_registered?
```

Comment:
What do you mean? we have specs to test against that. :)

The problem is that when an mhv account is initialized, it initially becomes "eligible", and no longer has the state of being upgraded that we explicitly set when we determined that the its only been registered and somehow became 'Premium' on its own. Could you elaborate on what you mean here? Note: i did have to change the logic around a bit for some of those "previousl_" methods.

Reply:
So the `previously_registered_somehow_upgraded?` logic is:

```
     previously_registered? && changes[:account_state]&.first == 'upgraded' 
```

This certainly covers the historical records that are upgraded with a registered timestamp but no upgraded timestamp.

I think this would also cover the happy path for accounts that we've created and upgraded. `previously_registered_somehow_upgraded?` would return true for them too. Those accounts will have been persisted with the `upgraded` state, then turn `eligible` in the eligibility check, and then will turn back to `upgraded` here because their initial states were `upgraded`.

But I guess this condition doesn't capture accounts that were existing and we upgraded because they don't have registered timestamps. I just wanted to point out that this transition seems to include accounts that we've created and upgraded.

Reply:
Yes that's by design. We only want it to capture accounts that we've registered, that we have not upgraded, but that may already be marked as "upgraded" because that's how we used to do things. 

I think in a future iteration we should consider possibilities where an account level could go down from Premium to Advanced as well.

Reply:
Right, but my point is that it's not _only_ that case. It also applies to accounts that we've registered and that we've upgraded.

Reply:
I think the logic for previously_registered is making sure there is no upgraded_at, no?